<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dipayan Akhuli">

<title>Exercise 03</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="sta426-week3-ex_files/libs/clipboard/clipboard.min.js"></script>
<script src="sta426-week3-ex_files/libs/quarto-html/quarto.js"></script>
<script src="sta426-week3-ex_files/libs/quarto-html/popper.min.js"></script>
<script src="sta426-week3-ex_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="sta426-week3-ex_files/libs/quarto-html/anchor.min.js"></script>
<link href="sta426-week3-ex_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="sta426-week3-ex_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="sta426-week3-ex_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="sta426-week3-ex_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="sta426-week3-ex_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Exercise 03</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Dipayan Akhuli </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="exploratory-data-analysis" class="level2">
<h2 class="anchored" data-anchor-id="exploratory-data-analysis">Exploratory Data Analysis</h2>
<p>Do an exploratory data analysis of a matrix of expression values. The data consists of expression values for samples that were treated with DMSO and TSA. The samples were measured using three technologies: bulk, IFC96, IFC800. See the two RDS files <code>counts.RDS</code> and <code>phenodata.RDS</code>.</p>
<p>The experiment has two experimental factors “Treatment” and “Technology”. Treatment has two levels, Technology has three levels. Samples that have the same combination of Treatment &amp; Technology are replicates.</p>
</section>
<section id="data-import" class="level2">
<h2 class="anchored" data-anchor-id="data-import">Data Import</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"counts.RDS"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>anno <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"phenodata.RDS"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(anno)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      Treatment Technology
bulk.DMSO-Pre-96IFC        DMSO       bulk
bulk.TSA-Pre-96IFC          TSA       bulk
bulk.DMSO-Pre-800IFC       DMSO       bulk
bulk.TSA-Pre-800IFC         TSA       bulk
bulk.DMSO-Post-800IFC      DMSO       bulk
bulk.TSA-Post-800IFC        TSA       bulk</code></pre>
</div>
</div>
</section>
<section id="compute-and-visualize-basic-statistics-on-the-experimental-design" class="level2">
<h2 class="anchored" data-anchor-id="compute-and-visualize-basic-statistics-on-the-experimental-design">Compute and visualize basic statistics on the experimental design</h2>
<p>How many replicates are there for each combination of factor levels?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># repls stores number of replicates per sample</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>repls <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Treatment)), <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Technology))))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(repls) <span class="ot">=</span> <span class="fu">unique</span>(anno<span class="sc">$</span>Treatment)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(repls) <span class="ot">=</span> <span class="fu">unique</span>(anno<span class="sc">$</span>Technology)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anno)) {</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  repls[anno<span class="sc">$</span>Treatment[i], anno<span class="sc">$</span>Technology[i]] <span class="ot">&lt;-</span> repls[anno<span class="sc">$</span>Treatment[i], anno<span class="sc">$</span>Technology[i]] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>repls</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     bulk IFC96 IFC800
DMSO    3    20     20
TSA     3    20     20</code></pre>
</div>
</div>
<p>How many genes have an expression value above 0 in each sample?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># num_genes stores number of expressed genes (expression value &gt; 0) per sample</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>num_genes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colSums</span>(x <span class="sc">!=</span> <span class="dv">0</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(num_genes) <span class="ot">=</span> <span class="st">"num_genes"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(num_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      num_genes
bulk.DMSO-Pre-96IFC       15163
bulk.TSA-Pre-96IFC        15261
bulk.DMSO-Pre-800IFC      15404
bulk.TSA-Pre-800IFC       16182
bulk.DMSO-Post-800IFC     15202
bulk.TSA-Post-800IFC      16167</code></pre>
</div>
</div>
<p>Are there systematic differences between the samples in the different conditions (treatment &amp; technology)? Visualize the following statistics per sample.</p>
<section id="number-of-detected-genes" class="level3">
<h3 class="anchored" data-anchor-id="number-of-detected-genes">Number of detected genes</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">cbind</span>(anno, num_genes), <span class="fu">aes</span>(<span class="at">x =</span> Technology, <span class="at">y =</span> num_genes, <span class="at">fill =</span> Treatment)) <span class="sc">+</span> </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">trim =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Samples"</span>, <span class="at">y =</span> <span class="st">"Number of detected genes"</span>) <span class="sc">+</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">+</span><span class="dv">3</span>),</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>, <span class="at">lineend =</span> <span class="st">"square"</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sta426-week3-ex_files/figure-html/viol-num-genes-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The number of detected genes is much higher in the bulk samples as compared to the IFC800 and IFC96 samples. Within the IFC800 and IFC96 samples, the TSA treated samples have a much larger spread as compared to the DMSO treated samples. Also, the mean is slightly higher for the IFC96 samples than the IFC800 samples. However, for each of the 3 technologies, the mean number of detected genes for both treatments is quite similar.</p>
</section>
<section id="total-sum-of-counts-per-sample-serves-as-a-proxy-for-the-number-of-sequenced-reads-per-sample" class="level3">
<h3 class="anchored" data-anchor-id="total-sum-of-counts-per-sample-serves-as-a-proxy-for-the-number-of-sequenced-reads-per-sample">Total sum of counts per sample (serves as a proxy for the number of sequenced reads per sample)</h3>
<p>Color the samples by the experimental condition.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>sum_counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colSums</span>(x))</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sum_counts) <span class="ot">=</span> <span class="st">"sum_counts"</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sum_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                      sum_counts
bulk.DMSO-Pre-96IFC     16469072
bulk.TSA-Pre-96IFC      10899777
bulk.DMSO-Pre-800IFC    22928016
bulk.TSA-Pre-800IFC     21170430
bulk.DMSO-Post-800IFC   16572277
bulk.TSA-Post-800IFC    21218842</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">cbind</span>(anno, sum_counts), <span class="fu">aes</span>(<span class="at">x =</span> Technology, <span class="at">y =</span> sum_counts, <span class="at">fill =</span> Treatment)) <span class="sc">+</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">trim =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>),</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">trans =</span> <span class="fu">log10_trans</span>()</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Samples"</span>, <span class="at">y =</span> <span class="st">"Sum of counts"</span>) <span class="sc">+</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">+</span><span class="dv">3</span>),</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>, <span class="at">lineend =</span> <span class="st">"square"</span>)</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sta426-week3-ex_files/figure-html/viol-sum-counts-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The total sum of counts for the bulk samples is 1-2 orders of magnitude greater than the IFC800 and IFC96 samples. Again, the mean is slightly higher for the IFC96 samples than the IFC800 samples, and for each of the 3 technologies, the total sum of counts for both treatments is quite similar.</p>
</section>
</section>
<section id="normalize-the-data" class="level2">
<h2 class="anchored" data-anchor-id="normalize-the-data">Normalize the data</h2>
<p>Scale the columns so that the total sum of all columns are identical.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># norm_x scales the columns such that each column sum is equal to the maximum of the column sums of x</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>norm_x <span class="ot">&lt;-</span> <span class="fu">sweep</span>(x<span class="sc">*</span><span class="fu">max</span>(sum_counts), <span class="dv">2</span>, <span class="fu">t</span>(sum_counts), <span class="st">"/"</span>)         </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="transform-the-data-to-log-scale" class="level2">
<h2 class="anchored" data-anchor-id="transform-the-data-to-log-scale">Transform the data to log-scale</h2>
<p>Use the function <code>log1p</code> to transform the data to log-scale.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># log_x transforms the normalized data to log-scale</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>log_x <span class="ot">&lt;-</span> <span class="fu">log1p</span>(norm_x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="visualize-the-distribution-of-the-expression-values" class="level2">
<h2 class="anchored" data-anchor-id="visualize-the-distribution-of-the-expression-values">Visualize the distribution of the expression values</h2>
<p>Use violin plots and boxplots to visualize the distribution of the expression values. Aggregate the replicates in the same experimental group into one average sample, and plot the averaged sample. Color by experimental factors.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># aggr_x contains expression data of all genes averaged over replicates for each sample</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>aggr_x <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Treatment))<span class="sc">*</span><span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Technology)), <span class="fu">nrow</span>(log_x)))</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(aggr_x) <span class="ot">=</span> <span class="fu">rownames</span>(log_x)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>cats <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Treatment =</span> <span class="fu">rep</span>(<span class="fu">rownames</span>(repls), <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Technology))), <span class="at">Technology =</span> <span class="fu">rep</span>(<span class="fu">colnames</span>(repls), <span class="at">each =</span> <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Treatment))))</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>aggr_x <span class="ot">&lt;-</span> <span class="fu">cbind</span>(cats, aggr_x)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(aggr_x)) {</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>aggr_x[i, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(aggr_x)] <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(log_x[, anno<span class="sc">$</span>Treatment <span class="sc">==</span> aggr_x<span class="sc">$</span>Treatment[i] <span class="sc">&amp;</span> anno<span class="sc">$</span>Technology <span class="sc">==</span> aggr_x<span class="sc">$</span>Technology[i]])</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># grp_x reshapes aggr_x for easy grouped violin plot</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>grp_x <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">nrow</span>(log_x)<span class="sc">*</span><span class="fu">nrow</span>(aggr_x), <span class="dv">3</span>))</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(grp_x) <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">colnames</span>(aggr_x)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="st">"expr"</span>)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>grp_x<span class="sc">$</span>Treatment <span class="ot">&lt;-</span> <span class="fu">rep</span>(aggr_x<span class="sc">$</span>Treatment, <span class="fu">nrow</span>(log_x))</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>grp_x<span class="sc">$</span>Technology <span class="ot">&lt;-</span> <span class="fu">rep</span>(aggr_x<span class="sc">$</span>Technology, <span class="fu">nrow</span>(log_x))</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>grp_x<span class="sc">$</span>expr <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">unname</span>(<span class="fu">as.matrix</span>(aggr_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(aggr_x)])))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(grp_x, <span class="fu">aes</span>(<span class="at">x =</span> Technology, <span class="at">y =</span> expr, <span class="at">fill =</span> Treatment)) <span class="sc">+</span> </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> .<span class="dv">05</span>, <span class="at">outlier.colour =</span> <span class="cn">NA</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.9</span>)) <span class="sc">+</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Samples"</span>, <span class="at">y =</span> <span class="st">"Expression values (scaled and transformed)"</span>) <span class="sc">+</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">+</span><span class="dv">3</span>),</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>, <span class="at">lineend =</span> <span class="st">"square"</span>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sta426-week3-ex_files/figure-html/viol-expr-vals-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>For all 3 technologies, the violin plots of the expression values look very similar for both treatments. For IFC800 samples, most expression values are very low, while IFC96 samples have slightly higher expression values. However, most strikingly, bulk samples have a considerably high fraction of higher expression values out of all expression values as compared to the IFC800 and IFC96 samples.</p>
</section>
<section id="most-variable-genes" class="level2">
<h2 class="anchored" data-anchor-id="most-variable-genes">Most variable genes</h2>
<p>Identify the 500 most variable genes (with largest variance across samples) and continue working with those.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># expr_vars contains variances in expression across samples for each gene</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>expr_vars <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colVars</span>(<span class="fu">as.matrix</span>(aggr_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(aggr_x)])))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(expr_vars) <span class="ot">=</span> <span class="st">"expr_variance"</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co"># top_vars sorts expr_vars in descending order and stores the 500 most variable genes</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>top_vars <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="sc">-</span><span class="fu">as.matrix</span>(expr_vars))[<span class="dv">1</span><span class="sc">:</span><span class="dv">500</span>]</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">#top_x contains expression data of 500 most variable genes</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>top_x <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">subset</span>(aggr_x, <span class="at">select =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)), <span class="fu">subset</span>(aggr_x, <span class="at">select =</span> <span class="dv">2</span> <span class="sc">+</span> top_vars))</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(top_x) <span class="ot">=</span> <span class="fu">unname</span>(<span class="fu">paste0</span>(top_x<span class="sc">$</span>Technology, <span class="st">'-'</span>, top_x<span class="sc">$</span>Treatment))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sample-correlations" class="level2">
<h2 class="anchored" data-anchor-id="sample-correlations">Sample correlations</h2>
<p>Compute and visualize the sample-to-sample correlations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cor_x is obtained from top_x by subtracting mean expression of all genes from each expression value per sample</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>cor_x <span class="ot">&lt;-</span> <span class="fu">cbind</span>(top_x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">mean_expr =</span> <span class="fu">rowMeans</span>(top_x[<span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)]), top_x[<span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)])</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>cor_x[, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="ot">&lt;-</span> cor_x[, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">-</span> cor_x<span class="sc">$</span>mean_expr</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># correl stores Pearson correlation coefficient between each sample</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>correl <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">nrow</span>(top_x), <span class="fu">nrow</span>(top_x))</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(top_x)) {</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>i) {</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    numr <span class="ot">&lt;-</span> <span class="fu">sum</span>(cor_x[i, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">*</span> cor_x[j, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)])</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    denr <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">sum</span>(cor_x[i, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">^</span> <span class="dv">2</span>) <span class="sc">*</span> <span class="fu">sum</span>(cor_x[j, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">^</span> <span class="dv">2</span>))</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    correl[i,j] <span class="ot">&lt;-</span> numr<span class="sc">/</span>denr;</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="co"># correl is converted to a symmetric dataframe for visualization</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>correl <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(correl <span class="sc">+</span> <span class="fu">t</span>(correl) <span class="sc">-</span> <span class="fu">diag</span>(<span class="fu">diag</span>(correl)));</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(correl) <span class="ot">=</span> <span class="fu">unname</span>(<span class="fu">paste0</span>(top_x<span class="sc">$</span>Technology, <span class="st">'-'</span>, top_x<span class="sc">$</span>Treatment))</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(correl) <span class="ot">=</span> <span class="fu">colnames</span>(correl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>melted_correl <span class="ot">&lt;-</span> <span class="fu">melt</span>(<span class="fu">cbind</span>(<span class="at">sample =</span> <span class="fu">colnames</span>(correl), correl), <span class="at">id.vars =</span> <span class="st">"sample"</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(melted_correl, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(sample, <span class="at">level =</span> <span class="fu">rownames</span>(correl)), <span class="at">y =</span> variable, <span class="at">fill =</span> value)) <span class="sc">+</span> </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tile</span>(<span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">high =</span> <span class="st">"red"</span>, <span class="at">mid =</span> <span class="st">"white"</span>, </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">midpoint =</span> <span class="dv">0</span>, <span class="at">limit =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>), <span class="at">space =</span> <span class="st">"Lab"</span>, </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">name=</span> <span class="st">"Correlation"</span>) <span class="sc">+</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="fu">aes</span>(variable, sample, <span class="at">label =</span> <span class="fu">round</span>(value, <span class="dv">3</span>)), <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>),</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">11</span>),</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">size =</span> <span class="dv">9</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_fixed</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sta426-week3-ex_files/figure-html/correl-expr-vals-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>For the bulk and IFC96 samples, the expression profiles of the DMSO and TSA treated samples are highly correlated, while the correlation is quite less for the IFC800 samples. There is also a slight positive correlation between IFC800 samples and both bulk and IFC96 samples, and a slight negative correlation between bulk and IFC96 samples.</p>
</section>
<section id="clustering" class="level2">
<h2 class="anchored" data-anchor-id="clustering">Clustering</h2>
<p>Compute and visualize a hierarchical clustering of the samples, use the method <code>hclust</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># hc_x stores hierarchical clusters of samples calculated using Euclidean distance</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>hc_x <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(top_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">9</span>, <span class="fl">4.1</span>, <span class="fl">4.1</span>, <span class="fl">2.1</span>), <span class="at">cex =</span> <span class="fl">0.7</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">as.dendrogram</span>(hc_x), <span class="at">ylab =</span> <span class="st">""</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">side =</span> <span class="dv">1</span>, <span class="at">text =</span> <span class="st">"Samples"</span>, <span class="at">line =</span> <span class="dv">7</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">side =</span> <span class="dv">2</span>, <span class="at">text =</span> <span class="st">"Cluster distances"</span>, <span class="at">line =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sta426-week3-ex_files/figure-html/hclust-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="heatmap" class="level2">
<h2 class="anchored" data-anchor-id="heatmap">Heatmap</h2>
<p>Use the package <code>pheatmap</code> to generate a heatmap of the expression data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap of normalized and transformed expression values of all genes per sample</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># genes are also clustered hierarchically based on Euclidean distance between expression values  </span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: heatmap</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 20</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">t</span>(<span class="fu">as.matrix</span>(top_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)])), <span class="at">cluster_rows =</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(<span class="fu">t</span>(top_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)]))), <span class="at">cluster_cols =</span> hc_x, <span class="at">show_rownames =</span> <span class="cn">FALSE</span>, <span class="at">show_colnames =</span> <span class="cn">TRUE</span>, <span class="at">angle_col =</span> <span class="dv">0</span>, <span class="at">cellwidth =</span> <span class="dv">65</span>, <span class="at">cellheight =</span> <span class="fl">0.8</span>, <span class="at">fontsize_col =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="sta426-week3-ex_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The most visually striking observation is that the bulk samples have higher expression values than the IFC800 and IFC96 samples, except for one specific cluster of genes which has low expression, and is expressed relatively more in the IFC96 samples. Expression values for the IFC800 samples are generally low, while those for IFC96 are also mostly low, except for a few clusters of genes which are similar to the bulk samples.</p>
<p>Coming to the effect of treatment, there are very few differences between DMSO and TSA treatment for the bulk samples. For the IFC800 samples, there is one small cluster of genes which has somewhat lower expression in the TSA treatment as compared to the DMSO treatment. For the IFC96 samples, there seem to be a few clusters of genes which have lower expression for the TSA treated samples as compared to the DMSO treated samples, while there is one cluster which has slightly higher expression.</p>
<!-- -->

</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb21" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Exercise 03"</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Dipayan Akhuli"</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: false</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="in">```{r setup, include=FALSE}</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">echo =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-packages</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(matrixStats)</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2)</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exploratory Data Analysis</span></span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>Do an exploratory data analysis of a matrix of expression values. The data consists of expression values for samples that were treated with DMSO and TSA. The samples were measured using three technologies: bulk, IFC96, IFC800. See the two RDS files <span class="in">`counts.RDS`</span> and <span class="in">`phenodata.RDS`</span>.</span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a>The experiment has two experimental factors "Treatment" and "Technology". Treatment has two levels, Technology has three levels. Samples that have the same combination of Treatment &amp; Technology are replicates.</span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a><span class="fu">## Data Import</span></span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a>x <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"counts.RDS"</span>)</span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a>anno <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"phenodata.RDS"</span>)</span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(anno)</span>
<span id="cb21-42"><a href="#cb21-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-43"><a href="#cb21-43" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-44"><a href="#cb21-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-45"><a href="#cb21-45" aria-hidden="true" tabindex="-1"></a><span class="fu">## Compute and visualize basic statistics on the experimental design</span></span>
<span id="cb21-46"><a href="#cb21-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-47"><a href="#cb21-47" aria-hidden="true" tabindex="-1"></a>How many replicates are there for each combination of factor levels?</span>
<span id="cb21-48"><a href="#cb21-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-51"><a href="#cb21-51" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-52"><a href="#cb21-52" aria-hidden="true" tabindex="-1"></a><span class="co"># repls stores number of replicates per sample</span></span>
<span id="cb21-53"><a href="#cb21-53" aria-hidden="true" tabindex="-1"></a>repls <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Treatment)), <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Technology))))</span>
<span id="cb21-54"><a href="#cb21-54" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(repls) <span class="ot">=</span> <span class="fu">unique</span>(anno<span class="sc">$</span>Treatment)</span>
<span id="cb21-55"><a href="#cb21-55" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(repls) <span class="ot">=</span> <span class="fu">unique</span>(anno<span class="sc">$</span>Technology)</span>
<span id="cb21-56"><a href="#cb21-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-57"><a href="#cb21-57" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(anno)) {</span>
<span id="cb21-58"><a href="#cb21-58" aria-hidden="true" tabindex="-1"></a>  repls[anno<span class="sc">$</span>Treatment[i], anno<span class="sc">$</span>Technology[i]] <span class="ot">&lt;-</span> repls[anno<span class="sc">$</span>Treatment[i], anno<span class="sc">$</span>Technology[i]] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb21-59"><a href="#cb21-59" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-60"><a href="#cb21-60" aria-hidden="true" tabindex="-1"></a>repls</span>
<span id="cb21-61"><a href="#cb21-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-62"><a href="#cb21-62" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-63"><a href="#cb21-63" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-64"><a href="#cb21-64" aria-hidden="true" tabindex="-1"></a>How many genes have an expression value above 0 in each sample?</span>
<span id="cb21-65"><a href="#cb21-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-68"><a href="#cb21-68" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-69"><a href="#cb21-69" aria-hidden="true" tabindex="-1"></a><span class="co"># num_genes stores number of expressed genes (expression value &gt; 0) per sample</span></span>
<span id="cb21-70"><a href="#cb21-70" aria-hidden="true" tabindex="-1"></a>num_genes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colSums</span>(x <span class="sc">!=</span> <span class="dv">0</span>))</span>
<span id="cb21-71"><a href="#cb21-71" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(num_genes) <span class="ot">=</span> <span class="st">"num_genes"</span></span>
<span id="cb21-72"><a href="#cb21-72" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(num_genes)</span>
<span id="cb21-73"><a href="#cb21-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-74"><a href="#cb21-74" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-75"><a href="#cb21-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-76"><a href="#cb21-76" aria-hidden="true" tabindex="-1"></a>Are there systematic differences between the samples in the different conditions (treatment &amp; technology)? Visualize the following statistics per sample.</span>
<span id="cb21-77"><a href="#cb21-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-78"><a href="#cb21-78" aria-hidden="true" tabindex="-1"></a><span class="fu">### Number of detected genes</span></span>
<span id="cb21-79"><a href="#cb21-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-82"><a href="#cb21-82" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-83"><a href="#cb21-83" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: viol-num-genes</span></span>
<span id="cb21-84"><a href="#cb21-84" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">cbind</span>(anno, num_genes), <span class="fu">aes</span>(<span class="at">x =</span> Technology, <span class="at">y =</span> num_genes, <span class="at">fill =</span> Treatment)) <span class="sc">+</span> </span>
<span id="cb21-85"><a href="#cb21-85" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">trim =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb21-86"><a href="#cb21-86" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Samples"</span>, <span class="at">y =</span> <span class="st">"Number of detected genes"</span>) <span class="sc">+</span></span>
<span id="cb21-87"><a href="#cb21-87" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb21-88"><a href="#cb21-88" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb21-89"><a href="#cb21-89" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb21-90"><a href="#cb21-90" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">+</span><span class="dv">3</span>),</span>
<span id="cb21-91"><a href="#cb21-91" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-92"><a href="#cb21-92" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>, <span class="at">lineend =</span> <span class="st">"square"</span>)</span>
<span id="cb21-93"><a href="#cb21-93" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-94"><a href="#cb21-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-95"><a href="#cb21-95" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-96"><a href="#cb21-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-97"><a href="#cb21-97" aria-hidden="true" tabindex="-1"></a>The number of detected genes is much higher in the bulk samples as compared to the IFC800 and IFC96 samples. Within the IFC800 and IFC96 samples, the TSA treated samples have a much larger spread as compared to the DMSO treated samples. Also, the mean is slightly higher for the IFC96 samples than the IFC800 samples. However, for each of the 3 technologies, the mean number of detected genes for both treatments is quite similar.</span>
<span id="cb21-98"><a href="#cb21-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-99"><a href="#cb21-99" aria-hidden="true" tabindex="-1"></a><span class="fu">### Total sum of counts per sample (serves as a proxy for the number of sequenced reads per sample)</span></span>
<span id="cb21-100"><a href="#cb21-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-101"><a href="#cb21-101" aria-hidden="true" tabindex="-1"></a>Color the samples by the experimental condition.</span>
<span id="cb21-102"><a href="#cb21-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-105"><a href="#cb21-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-106"><a href="#cb21-106" aria-hidden="true" tabindex="-1"></a>sum_counts <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colSums</span>(x))</span>
<span id="cb21-107"><a href="#cb21-107" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sum_counts) <span class="ot">=</span> <span class="st">"sum_counts"</span></span>
<span id="cb21-108"><a href="#cb21-108" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sum_counts)</span>
<span id="cb21-109"><a href="#cb21-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-110"><a href="#cb21-110" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-111"><a href="#cb21-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-114"><a href="#cb21-114" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-115"><a href="#cb21-115" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: viol-sum-counts</span></span>
<span id="cb21-116"><a href="#cb21-116" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">cbind</span>(anno, sum_counts), <span class="fu">aes</span>(<span class="at">x =</span> Technology, <span class="at">y =</span> sum_counts, <span class="at">fill =</span> Treatment)) <span class="sc">+</span> </span>
<span id="cb21-117"><a href="#cb21-117" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">trim =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb21-118"><a href="#cb21-118" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(</span>
<span id="cb21-119"><a href="#cb21-119" aria-hidden="true" tabindex="-1"></a>    <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>),</span>
<span id="cb21-120"><a href="#cb21-120" aria-hidden="true" tabindex="-1"></a>    <span class="at">trans =</span> <span class="fu">log10_trans</span>()</span>
<span id="cb21-121"><a href="#cb21-121" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb21-122"><a href="#cb21-122" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Samples"</span>, <span class="at">y =</span> <span class="st">"Sum of counts"</span>) <span class="sc">+</span></span>
<span id="cb21-123"><a href="#cb21-123" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb21-124"><a href="#cb21-124" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb21-125"><a href="#cb21-125" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb21-126"><a href="#cb21-126" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">+</span><span class="dv">3</span>),</span>
<span id="cb21-127"><a href="#cb21-127" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-128"><a href="#cb21-128" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>, <span class="at">lineend =</span> <span class="st">"square"</span>)</span>
<span id="cb21-129"><a href="#cb21-129" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-130"><a href="#cb21-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-131"><a href="#cb21-131" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-132"><a href="#cb21-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-133"><a href="#cb21-133" aria-hidden="true" tabindex="-1"></a>The total sum of counts for the bulk samples is 1-2 orders of magnitude greater than the IFC800 and IFC96 samples. Again, the mean is slightly higher for the IFC96 samples than the IFC800 samples, and for each of the 3 technologies, the total sum of counts for both treatments is quite similar.</span>
<span id="cb21-134"><a href="#cb21-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-135"><a href="#cb21-135" aria-hidden="true" tabindex="-1"></a><span class="fu">## Normalize the data</span></span>
<span id="cb21-136"><a href="#cb21-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-137"><a href="#cb21-137" aria-hidden="true" tabindex="-1"></a>Scale the columns so that the total sum of all columns are identical.</span>
<span id="cb21-138"><a href="#cb21-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-141"><a href="#cb21-141" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-142"><a href="#cb21-142" aria-hidden="true" tabindex="-1"></a><span class="co"># norm_x scales the columns such that each column sum is equal to the maximum of the column sums of x</span></span>
<span id="cb21-143"><a href="#cb21-143" aria-hidden="true" tabindex="-1"></a>norm_x <span class="ot">&lt;-</span> <span class="fu">sweep</span>(x<span class="sc">*</span><span class="fu">max</span>(sum_counts), <span class="dv">2</span>, <span class="fu">t</span>(sum_counts), <span class="st">"/"</span>)         </span>
<span id="cb21-144"><a href="#cb21-144" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-145"><a href="#cb21-145" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-146"><a href="#cb21-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-147"><a href="#cb21-147" aria-hidden="true" tabindex="-1"></a><span class="fu">## Transform the data to log-scale</span></span>
<span id="cb21-148"><a href="#cb21-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-149"><a href="#cb21-149" aria-hidden="true" tabindex="-1"></a>Use the function <span class="in">`log1p`</span> to transform the data to log-scale.</span>
<span id="cb21-150"><a href="#cb21-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-153"><a href="#cb21-153" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-154"><a href="#cb21-154" aria-hidden="true" tabindex="-1"></a><span class="co"># log_x transforms the normalized data to log-scale</span></span>
<span id="cb21-155"><a href="#cb21-155" aria-hidden="true" tabindex="-1"></a>log_x <span class="ot">&lt;-</span> <span class="fu">log1p</span>(norm_x)</span>
<span id="cb21-156"><a href="#cb21-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-157"><a href="#cb21-157" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-158"><a href="#cb21-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-159"><a href="#cb21-159" aria-hidden="true" tabindex="-1"></a><span class="fu">## Visualize the distribution of the expression values</span></span>
<span id="cb21-160"><a href="#cb21-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-161"><a href="#cb21-161" aria-hidden="true" tabindex="-1"></a>Use violin plots and boxplots to visualize the distribution of the expression values. Aggregate the replicates in the same experimental group into one average sample, and plot the averaged sample. Color by experimental factors.</span>
<span id="cb21-162"><a href="#cb21-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-165"><a href="#cb21-165" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-166"><a href="#cb21-166" aria-hidden="true" tabindex="-1"></a><span class="co"># aggr_x contains expression data of all genes averaged over replicates for each sample</span></span>
<span id="cb21-167"><a href="#cb21-167" aria-hidden="true" tabindex="-1"></a>aggr_x <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Treatment))<span class="sc">*</span><span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Technology)), <span class="fu">nrow</span>(log_x)))</span>
<span id="cb21-168"><a href="#cb21-168" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(aggr_x) <span class="ot">=</span> <span class="fu">rownames</span>(log_x)</span>
<span id="cb21-169"><a href="#cb21-169" aria-hidden="true" tabindex="-1"></a>cats <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Treatment =</span> <span class="fu">rep</span>(<span class="fu">rownames</span>(repls), <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Technology))), <span class="at">Technology =</span> <span class="fu">rep</span>(<span class="fu">colnames</span>(repls), <span class="at">each =</span> <span class="fu">length</span>(<span class="fu">unique</span>(anno<span class="sc">$</span>Treatment))))</span>
<span id="cb21-170"><a href="#cb21-170" aria-hidden="true" tabindex="-1"></a>aggr_x <span class="ot">&lt;-</span> <span class="fu">cbind</span>(cats, aggr_x)</span>
<span id="cb21-171"><a href="#cb21-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-172"><a href="#cb21-172" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(aggr_x)) {</span>
<span id="cb21-173"><a href="#cb21-173" aria-hidden="true" tabindex="-1"></a>aggr_x[i, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(aggr_x)] <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(log_x[, anno<span class="sc">$</span>Treatment <span class="sc">==</span> aggr_x<span class="sc">$</span>Treatment[i] <span class="sc">&amp;</span> anno<span class="sc">$</span>Technology <span class="sc">==</span> aggr_x<span class="sc">$</span>Technology[i]])</span>
<span id="cb21-174"><a href="#cb21-174" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-175"><a href="#cb21-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-176"><a href="#cb21-176" aria-hidden="true" tabindex="-1"></a><span class="co"># grp_x reshapes aggr_x for easy grouped violin plot</span></span>
<span id="cb21-177"><a href="#cb21-177" aria-hidden="true" tabindex="-1"></a>grp_x <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">nrow</span>(log_x)<span class="sc">*</span><span class="fu">nrow</span>(aggr_x), <span class="dv">3</span>))</span>
<span id="cb21-178"><a href="#cb21-178" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(grp_x) <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">colnames</span>(aggr_x)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="st">"expr"</span>)</span>
<span id="cb21-179"><a href="#cb21-179" aria-hidden="true" tabindex="-1"></a>grp_x<span class="sc">$</span>Treatment <span class="ot">&lt;-</span> <span class="fu">rep</span>(aggr_x<span class="sc">$</span>Treatment, <span class="fu">nrow</span>(log_x))</span>
<span id="cb21-180"><a href="#cb21-180" aria-hidden="true" tabindex="-1"></a>grp_x<span class="sc">$</span>Technology <span class="ot">&lt;-</span> <span class="fu">rep</span>(aggr_x<span class="sc">$</span>Technology, <span class="fu">nrow</span>(log_x))</span>
<span id="cb21-181"><a href="#cb21-181" aria-hidden="true" tabindex="-1"></a>grp_x<span class="sc">$</span>expr <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">unname</span>(<span class="fu">as.matrix</span>(aggr_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(aggr_x)])))</span>
<span id="cb21-182"><a href="#cb21-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-183"><a href="#cb21-183" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-184"><a href="#cb21-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-187"><a href="#cb21-187" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-188"><a href="#cb21-188" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: viol-expr-vals</span></span>
<span id="cb21-189"><a href="#cb21-189" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(grp_x, <span class="fu">aes</span>(<span class="at">x =</span> Technology, <span class="at">y =</span> expr, <span class="at">fill =</span> Treatment)) <span class="sc">+</span> </span>
<span id="cb21-190"><a href="#cb21-190" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>() <span class="sc">+</span></span>
<span id="cb21-191"><a href="#cb21-191" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> .<span class="dv">05</span>, <span class="at">outlier.colour =</span> <span class="cn">NA</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width =</span> <span class="fl">0.9</span>)) <span class="sc">+</span></span>
<span id="cb21-192"><a href="#cb21-192" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Samples"</span>, <span class="at">y =</span> <span class="st">"Expression values (scaled and transformed)"</span>) <span class="sc">+</span></span>
<span id="cb21-193"><a href="#cb21-193" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb21-194"><a href="#cb21-194" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb21-195"><a href="#cb21-195" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb21-196"><a href="#cb21-196" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">vjust =</span> <span class="sc">+</span><span class="dv">3</span>),</span>
<span id="cb21-197"><a href="#cb21-197" aria-hidden="true" tabindex="-1"></a>    <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-198"><a href="#cb21-198" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">color =</span> <span class="st">"black"</span>, <span class="at">linewidth =</span> <span class="fl">0.5</span>, <span class="at">lineend =</span> <span class="st">"square"</span>)</span>
<span id="cb21-199"><a href="#cb21-199" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-200"><a href="#cb21-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-201"><a href="#cb21-201" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-202"><a href="#cb21-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-203"><a href="#cb21-203" aria-hidden="true" tabindex="-1"></a>For all 3 technologies, the violin plots of the expression values look very similar for both treatments. For IFC800 samples, most expression values are very low, while IFC96 samples have slightly higher expression values. However, most strikingly, bulk samples have a considerably high fraction of higher expression values out of all expression values as compared to the IFC800 and IFC96 samples.</span>
<span id="cb21-204"><a href="#cb21-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-205"><a href="#cb21-205" aria-hidden="true" tabindex="-1"></a><span class="fu">## Most variable genes</span></span>
<span id="cb21-206"><a href="#cb21-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-207"><a href="#cb21-207" aria-hidden="true" tabindex="-1"></a>Identify the 500 most variable genes (with largest variance across samples) and continue working with those.</span>
<span id="cb21-208"><a href="#cb21-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-211"><a href="#cb21-211" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-212"><a href="#cb21-212" aria-hidden="true" tabindex="-1"></a><span class="co"># expr_vars contains variances in expression across samples for each gene</span></span>
<span id="cb21-213"><a href="#cb21-213" aria-hidden="true" tabindex="-1"></a>expr_vars <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colVars</span>(<span class="fu">as.matrix</span>(aggr_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(aggr_x)])))</span>
<span id="cb21-214"><a href="#cb21-214" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(expr_vars) <span class="ot">=</span> <span class="st">"expr_variance"</span></span>
<span id="cb21-215"><a href="#cb21-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-216"><a href="#cb21-216" aria-hidden="true" tabindex="-1"></a><span class="co"># top_vars sorts expr_vars in descending order and stores the 500 most variable genes</span></span>
<span id="cb21-217"><a href="#cb21-217" aria-hidden="true" tabindex="-1"></a>top_vars <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="sc">-</span><span class="fu">as.matrix</span>(expr_vars))[<span class="dv">1</span><span class="sc">:</span><span class="dv">500</span>]</span>
<span id="cb21-218"><a href="#cb21-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-219"><a href="#cb21-219" aria-hidden="true" tabindex="-1"></a><span class="co">#top_x contains expression data of 500 most variable genes</span></span>
<span id="cb21-220"><a href="#cb21-220" aria-hidden="true" tabindex="-1"></a>top_x <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">subset</span>(aggr_x, <span class="at">select =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)), <span class="fu">subset</span>(aggr_x, <span class="at">select =</span> <span class="dv">2</span> <span class="sc">+</span> top_vars))</span>
<span id="cb21-221"><a href="#cb21-221" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(top_x) <span class="ot">=</span> <span class="fu">unname</span>(<span class="fu">paste0</span>(top_x<span class="sc">$</span>Technology, <span class="st">'-'</span>, top_x<span class="sc">$</span>Treatment))</span>
<span id="cb21-222"><a href="#cb21-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-223"><a href="#cb21-223" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-224"><a href="#cb21-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-225"><a href="#cb21-225" aria-hidden="true" tabindex="-1"></a><span class="fu">## Sample correlations</span></span>
<span id="cb21-226"><a href="#cb21-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-227"><a href="#cb21-227" aria-hidden="true" tabindex="-1"></a>Compute and visualize the sample-to-sample correlations.</span>
<span id="cb21-228"><a href="#cb21-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-231"><a href="#cb21-231" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-232"><a href="#cb21-232" aria-hidden="true" tabindex="-1"></a><span class="co"># cor_x is obtained from top_x by subtracting mean expression of all genes from each expression value per sample</span></span>
<span id="cb21-233"><a href="#cb21-233" aria-hidden="true" tabindex="-1"></a>cor_x <span class="ot">&lt;-</span> <span class="fu">cbind</span>(top_x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">mean_expr =</span> <span class="fu">rowMeans</span>(top_x[<span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)]), top_x[<span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)])</span>
<span id="cb21-234"><a href="#cb21-234" aria-hidden="true" tabindex="-1"></a>cor_x[, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="ot">&lt;-</span> cor_x[, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">-</span> cor_x<span class="sc">$</span>mean_expr</span>
<span id="cb21-235"><a href="#cb21-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-236"><a href="#cb21-236" aria-hidden="true" tabindex="-1"></a><span class="co"># correl stores Pearson correlation coefficient between each sample</span></span>
<span id="cb21-237"><a href="#cb21-237" aria-hidden="true" tabindex="-1"></a>correl <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="fu">nrow</span>(top_x), <span class="fu">nrow</span>(top_x))</span>
<span id="cb21-238"><a href="#cb21-238" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(top_x)) {</span>
<span id="cb21-239"><a href="#cb21-239" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>i) {</span>
<span id="cb21-240"><a href="#cb21-240" aria-hidden="true" tabindex="-1"></a>    numr <span class="ot">&lt;-</span> <span class="fu">sum</span>(cor_x[i, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">*</span> cor_x[j, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)])</span>
<span id="cb21-241"><a href="#cb21-241" aria-hidden="true" tabindex="-1"></a>    denr <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">sum</span>(cor_x[i, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">^</span> <span class="dv">2</span>) <span class="sc">*</span> <span class="fu">sum</span>(cor_x[j, <span class="dv">4</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)] <span class="sc">^</span> <span class="dv">2</span>))</span>
<span id="cb21-242"><a href="#cb21-242" aria-hidden="true" tabindex="-1"></a>    correl[i,j] <span class="ot">&lt;-</span> numr<span class="sc">/</span>denr;</span>
<span id="cb21-243"><a href="#cb21-243" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-244"><a href="#cb21-244" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-245"><a href="#cb21-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-246"><a href="#cb21-246" aria-hidden="true" tabindex="-1"></a><span class="co"># correl is converted to a symmetric dataframe for visualization</span></span>
<span id="cb21-247"><a href="#cb21-247" aria-hidden="true" tabindex="-1"></a>correl <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(correl <span class="sc">+</span> <span class="fu">t</span>(correl) <span class="sc">-</span> <span class="fu">diag</span>(<span class="fu">diag</span>(correl)));</span>
<span id="cb21-248"><a href="#cb21-248" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(correl) <span class="ot">=</span> <span class="fu">unname</span>(<span class="fu">paste0</span>(top_x<span class="sc">$</span>Technology, <span class="st">'-'</span>, top_x<span class="sc">$</span>Treatment))</span>
<span id="cb21-249"><a href="#cb21-249" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(correl) <span class="ot">=</span> <span class="fu">colnames</span>(correl)</span>
<span id="cb21-250"><a href="#cb21-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-251"><a href="#cb21-251" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-252"><a href="#cb21-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-255"><a href="#cb21-255" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-256"><a href="#cb21-256" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: correl-expr-vals</span></span>
<span id="cb21-257"><a href="#cb21-257" aria-hidden="true" tabindex="-1"></a>melted_correl <span class="ot">&lt;-</span> <span class="fu">melt</span>(<span class="fu">cbind</span>(<span class="at">sample =</span> <span class="fu">colnames</span>(correl), correl), <span class="at">id.vars =</span> <span class="st">"sample"</span>)</span>
<span id="cb21-258"><a href="#cb21-258" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(melted_correl, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(sample, <span class="at">level =</span> <span class="fu">rownames</span>(correl)), <span class="at">y =</span> variable, <span class="at">fill =</span> value)) <span class="sc">+</span> </span>
<span id="cb21-259"><a href="#cb21-259" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tile</span>(<span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb21-260"><a href="#cb21-260" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">high =</span> <span class="st">"red"</span>, <span class="at">mid =</span> <span class="st">"white"</span>, </span>
<span id="cb21-261"><a href="#cb21-261" aria-hidden="true" tabindex="-1"></a>    <span class="at">midpoint =</span> <span class="dv">0</span>, <span class="at">limit =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>), <span class="at">space =</span> <span class="st">"Lab"</span>, </span>
<span id="cb21-262"><a href="#cb21-262" aria-hidden="true" tabindex="-1"></a>    <span class="at">name=</span> <span class="st">"Correlation"</span>) <span class="sc">+</span></span>
<span id="cb21-263"><a href="#cb21-263" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text</span>(<span class="fu">aes</span>(variable, sample, <span class="at">label =</span> <span class="fu">round</span>(value, <span class="dv">3</span>)), <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb21-264"><a href="#cb21-264" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb21-265"><a href="#cb21-265" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb21-266"><a href="#cb21-266" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>),</span>
<span id="cb21-267"><a href="#cb21-267" aria-hidden="true" tabindex="-1"></a>      <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">11</span>),</span>
<span id="cb21-268"><a href="#cb21-268" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-269"><a href="#cb21-269" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-270"><a href="#cb21-270" aria-hidden="true" tabindex="-1"></a>      <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-271"><a href="#cb21-271" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-272"><a href="#cb21-272" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-273"><a href="#cb21-273" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb21-274"><a href="#cb21-274" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb21-275"><a href="#cb21-275" aria-hidden="true" tabindex="-1"></a>      <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">size =</span> <span class="dv">9</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb21-276"><a href="#cb21-276" aria-hidden="true" tabindex="-1"></a>    <span class="fu">coord_fixed</span>()</span>
<span id="cb21-277"><a href="#cb21-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-278"><a href="#cb21-278" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-279"><a href="#cb21-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-280"><a href="#cb21-280" aria-hidden="true" tabindex="-1"></a>For the bulk and IFC96 samples, the expression profiles of the DMSO and TSA treated samples are highly correlated, while the correlation is quite less for the IFC800 samples. There is also a slight positive correlation between IFC800 samples and both bulk and IFC96 samples, and a slight negative correlation between bulk and IFC96 samples.</span>
<span id="cb21-281"><a href="#cb21-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-282"><a href="#cb21-282" aria-hidden="true" tabindex="-1"></a><span class="fu">## Clustering</span></span>
<span id="cb21-283"><a href="#cb21-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-284"><a href="#cb21-284" aria-hidden="true" tabindex="-1"></a>Compute and visualize a hierarchical clustering of the samples, use the method <span class="in">`hclust`</span>.</span>
<span id="cb21-285"><a href="#cb21-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-288"><a href="#cb21-288" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-289"><a href="#cb21-289" aria-hidden="true" tabindex="-1"></a><span class="co"># hc_x stores hierarchical clusters of samples calculated using Euclidean distance</span></span>
<span id="cb21-290"><a href="#cb21-290" aria-hidden="true" tabindex="-1"></a>hc_x <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(top_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)]))</span>
<span id="cb21-291"><a href="#cb21-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-292"><a href="#cb21-292" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-293"><a href="#cb21-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-296"><a href="#cb21-296" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb21-297"><a href="#cb21-297" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: hclust</span></span>
<span id="cb21-298"><a href="#cb21-298" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">9</span>, <span class="fl">4.1</span>, <span class="fl">4.1</span>, <span class="fl">2.1</span>), <span class="at">cex =</span> <span class="fl">0.7</span>)</span>
<span id="cb21-299"><a href="#cb21-299" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">as.dendrogram</span>(hc_x), <span class="at">ylab =</span> <span class="st">""</span>)</span>
<span id="cb21-300"><a href="#cb21-300" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">side =</span> <span class="dv">1</span>, <span class="at">text =</span> <span class="st">"Samples"</span>, <span class="at">line =</span> <span class="dv">7</span>)</span>
<span id="cb21-301"><a href="#cb21-301" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">side =</span> <span class="dv">2</span>, <span class="at">text =</span> <span class="st">"Cluster distances"</span>, <span class="at">line =</span> <span class="dv">3</span>)</span>
<span id="cb21-302"><a href="#cb21-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-303"><a href="#cb21-303" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-304"><a href="#cb21-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-305"><a href="#cb21-305" aria-hidden="true" tabindex="-1"></a><span class="fu">## Heatmap</span></span>
<span id="cb21-306"><a href="#cb21-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-307"><a href="#cb21-307" aria-hidden="true" tabindex="-1"></a>Use the package <span class="in">`pheatmap`</span> to generate a heatmap of the expression data.</span>
<span id="cb21-308"><a href="#cb21-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-309"><a href="#cb21-309" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.height=7}</span></span>
<span id="cb21-310"><a href="#cb21-310" aria-hidden="true" tabindex="-1"></a><span class="co"># heatmap of normalized and transformed expression values of all genes per sample</span></span>
<span id="cb21-311"><a href="#cb21-311" aria-hidden="true" tabindex="-1"></a><span class="co"># genes are also clustered hierarchically based on Euclidean distance between expression values  </span></span>
<span id="cb21-312"><a href="#cb21-312" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: heatmap</span></span>
<span id="cb21-313"><a href="#cb21-313" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 20</span></span>
<span id="cb21-314"><a href="#cb21-314" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">t</span>(<span class="fu">as.matrix</span>(top_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)])), <span class="at">cluster_rows =</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(<span class="fu">t</span>(top_x[, <span class="dv">3</span><span class="sc">:</span><span class="fu">ncol</span>(top_x)]))), <span class="at">cluster_cols =</span> hc_x, <span class="at">show_rownames =</span> <span class="cn">FALSE</span>, <span class="at">show_colnames =</span> <span class="cn">TRUE</span>, <span class="at">angle_col =</span> <span class="dv">0</span>, <span class="at">cellwidth =</span> <span class="dv">65</span>, <span class="at">cellheight =</span> <span class="fl">0.8</span>, <span class="at">fontsize_col =</span> <span class="dv">8</span>)</span>
<span id="cb21-315"><a href="#cb21-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-316"><a href="#cb21-316" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb21-317"><a href="#cb21-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-318"><a href="#cb21-318" aria-hidden="true" tabindex="-1"></a>The most visually striking observation is that the bulk samples have higher expression values than the IFC800 and IFC96 samples, except for one specific cluster of genes which has low expression, and is expressed relatively more in the IFC96 samples. Expression values for the IFC800 samples are generally low, while those for IFC96 are also mostly low, except for a few clusters of genes which are similar to the bulk samples.</span>
<span id="cb21-319"><a href="#cb21-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-320"><a href="#cb21-320" aria-hidden="true" tabindex="-1"></a>Coming to the effect of treatment, there are very few differences between DMSO and TSA treatment for the bulk samples. For the IFC800 samples, there is one small cluster of genes which has somewhat lower expression in the TSA treatment as compared to the DMSO treatment. For the IFC96 samples, there seem to be a few clusters of genes which have lower expression for the TSA treated samples as compared to the DMSO treated samples, while there is one cluster which has slightly higher expression.</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>